### PR TITLE
RAB-927: Upsert running user on job save

### PR DESCRIPTION
**Pull request by ValentinMumble**
**Description (for Contributor and Core Developer)**

Why we added `changeTrackingPolicy: DEFERRED_EXPLICIT ` ? 
Because when we persist the user then flush, the job instance is saved. Problem is that the job instance is not validated yet. As by default the changeTrackingPolicy is DEFERRED_IMPLICIT doctrine save the entity as he detect that there is a change in the entity. 

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Review by StevenVAIDIE — APPROVED** [review]

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/UserManagement/UpsertRunningUser.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    public function execute(string $jobCode, array $userGroupCodes): void
+    {
+        $username = sprintf('%s%s', self::AUTOMATED_USER_PREFIX, $jobCode);
+        $allRoleCodes = array_map(static fn (RoleInterface $role) => $role->getRole(), $this->roleRepository->findAll());
+
+        $upsertUserCommand = UpsertUserCommand::job(
+            $username,
+            'fakepassword',
+            sprintf('%s@fake.com', $username),
```
Comment (grena) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/UserManagement/UpsertRunningUser.php:35]:
```suggestion
            sprintf('%s@example.com', $username),
```
As `fake.com` is a real domain name, maybe leading to real email addresses

File: src/Akeneo/UserManagement/Bundle/Security/JobUserProvider.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadUserByUsername($username)
+    {
+        $user = $this->userRepository->findOneByIdentifier($username);
+        if (!$user || $user->isApiUser()) {
+            throw new UserNotFoundException(sprintf('User with username "%s" does not exist.', $username));
```
Comment (grena) [inline src/Akeneo/UserManagement/Bundle/Security/JobUserProvider.php:28]: This exception text could be misleading I think if the use really exists but is an API User. 
```suggestion
            throw new UserNotFoundException(sprintf('Job user with username "%s" does not exist, or exists but as an API user.', $username));
```

File: src/Akeneo/UserManagement/Bundle/Security/JobUserProvider.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public function refreshUser(UserInterface $user)
+    {
+        if (!$this->supportsClass($user::class)) {
+            throw new UnsupportedUserException(sprintf('User object of class "%s" is not supported.', $user::class));
+        }
+
+        $reloadedUser = $this->userRepository->find($user->getId());
+        if (null === $reloadedUser || $reloadedUser->isApiUser()) {
+            throw new UserNotFoundException(sprintf('User with id %d not found', $user->getId()));
```
Comment (grena) [inline src/Akeneo/UserManagement/Bundle/Security/JobUserProvider.php:49]: Same problem to me, exception text could be misleading for debug.

File: src/Akeneo/UserManagement/Bundle/Security/UserApiProvider.php [hunk 4]
```diff
@@ -36,7 +36,7 @@ public function __construct(UserRepositoryInterface $userRepository)
     public function loadUserByUsername($username)
     {
         $user = $this->userRepository->findOneByIdentifier($username);
-        if (!$user) {
+        if (!$user || $user->isJobUser()) {
             throw new UsernameNotFoundException(sprintf('User with username "%s" does not exist.', $username));
```
Comment (grena) [inline src/Akeneo/UserManagement/Bundle/Security/UserApiProvider.php:40]: Same etc.

File: src/Akeneo/UserManagement/Bundle/Validator/Constraints/CreateUser.php [hunk 5]
```diff
@@ -13,7 +13,8 @@
  */
 class CreateUser extends Constraint
 {
-    public $errorSpaceInUsername = 'The username should not contain space character.';
+    public string $errorSpaceInUsername = 'The username should not contain space character.';
```
Comment (grena) [inline src/Akeneo/UserManagement/Bundle/Validator/Constraints/CreateUser.php:16]:
```suggestion
    public string $errorSpaceInUsername = 'Username cannot contain whitespace character.';
```

File: tests/back/UserManagement/Specification/Bundle/Security/JobUserProviderSpec.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        $julia->isApiUser()->willReturn(false);
+        $julia->isEnabled()->willReturn(true);
+        $userRepository->findOneByIdentifier('julia')->willReturn($julia);
+        $this->loadUserByUsername('julia')->shouldReturn($julia);
+    }
+
+    public function it_throws_an_exception_if_username_does_not_exist(UserRepositoryInterface $userRepository)
+    {
+        $userRepository->findOneByIdentifier('jean-pacôme')->willReturn(null);
```
Comment (grena) [inline tests/back/UserManagement/Specification/Bundle/Security/JobUserProviderSpec.php:30]: :laughing:

File: src/Akeneo/Tool/Component/Batch/Model/JobInstance.php [hunk 7]
```diff
@@ -61,11 +61,9 @@ class JobInstance
     /** @var array */
     protected $rawParameters = [];
 
-    /** @var bool */
-    protected $scheduled = false;
+    protected bool $scheduled = false;
 
-    /** @var string */
-    protected $automation;
+    protected ?array $automation = null;
```
Comment (grena) [inline src/Akeneo/Tool/Component/Batch/Model/JobInstance.php:66]: For my knowledge, where do we turn json into an array?

**Review by grena — APPROVED** [review]

File: src/Akeneo/Tool/Component/Batch/Model/JobInstance.php (hunk 7 above)
Comment (ValentinMumble) [reply src/Akeneo/Tool/Component/Batch/Model/JobInstance.php:66]: I believe Doctrine does it when having a JSON value https://github.com/doctrine/dbal/blob/118a360e9437e88d49024f36283c8bcbd76105f5/src/Types/JsonType.php#L60

**Pull request merged**